logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -0,0 +1,96 @@
+# encoding: utf-8
+require 'singleton'
+require "rubygems/package"
+
+module LogStash
+  module Plugins
+    class Registry
+
+      include Singleton
+
+      attr_reader :registry
+
+      def initialize
+        @registry = {}
+        @logger = Cabin::Channel.get(LogStash)
+      end
+
+      def lookup(type, plugin, &block)
+        plugin_path = "logstash/#{type}s/#{plugin}"
+
+        if is_installed?(type, plugin)
+          return registry[plugin_path] if is_registered?(plugin_path)
+          begin
+            require plugin_path
+            klass = load_plugin(type, plugin, &block)
+            register(plugin_path, klass)
+            return klass
+          rescue => e
+            # plugin loading error
+            @logger.debug("Problems loading an installed plugin with", :type => type, :name => plugin, :path => plugin_path, :error => e) if @logger.debug?
+            raise LoadError, "Problems loading an installed plugin with" # (TODO) Need to be added to I18n when agreed on message.
+          end
+        elsif is_defined?(type, plugin)
+          # The plugin was defined directly in the code, so there is no need to use the
+          # require way of loading classes
+          return registry[plugin_path] if is_registered?(plugin_path)
+          klass = load_plugin(type, plugin, &block)
+          register(plugin_path, klass)
+          return klass
+        else # Plugin requested not currently installed.
+          @logger.debug("Requested plugin is not currently installed or defined", :type => type, :name => plugin, :path => plugin_path) if @logger.debug?
+          raise LoadError, "Requested plugin is not currently installed or defined" # (TODO) Need to be added to I18n when agreed on message.
+        end
+      end
+
+      def register(path, plugin)
```

Comment:
I think there is an small confusion here, the last argument of register is the plugin class, not the plugin name. But yes, I do agree this signature could be improved :+1: